core/src/main/java/com/amplifyframework/analytics/Properties.java

```diff
@@ -15,34 +15,152 @@
 
 package com.amplifyframework.analytics;
 
-import com.amplifyframework.util.Immutable;
+import androidx.annotation.NonNull;
 
 import java.util.HashMap;
+import java.util.Iterator;
 import java.util.Map;
 
 /**
- * Wrapper for {@link Map} to restrict the types that may be passed in.
+ * Properties provides key/value pairs to represent attributes of an {@link AnalyticsEvent}. While
+ * different analytics services will have support for different datatypes, Properties provides a
+ * baseline of support for String, Boolean, Double, and Integer. If one of these aren't available in
+ * a given service, the plugin is expected to cast it to something supported. (e.g., convert Boolean
+ * true to the String "true")
+ *
+ * <pre>
+ *     Properties properties = Properties.builder()
+ *          .add("PostType", "UserImage")
+ *          .add("LikedUserID", 78219)
+ *          .add("FirstLike", true)
+ *          .build();
+ * </pre>
  */
-public class Properties {
+public final class Properties implements Iterable<Map.Entry<String, Property<?>>> {
+    private final Map<String, Property<?>> properties;
+
+    private Properties(Map<String, Property<?>> properties) {
+        this.properties = properties;
+    }
+
+    /**
+     * Iterator allows Properties to support foreach operations.
+     *
+     * <pre>
+     *      for (Map.Entry&lt;String, Property&lt;?&gt;&gt; entry : properties) {
+     *          String key = entry.getKey();
+     *          Property&lt;?&gt; value = entry.getValue();
+     *
+     *          // Do something with key and value
+     *      }
+     * </pre>
+     *
+     * @return An {@link Iterator} of the underlying {@link Map#entrySet()} to enable foreach access
+     */
+    @Override
+    @NonNull
+    public Iterator<Map.Entry<String, Property<?>>> iterator() {
+        return properties.entrySet().iterator();
+    }
+
+    /**
+     * size returns the number of properties.
+     *
+     * @return The number of properties
+     */
+    @NonNull
+    public Integer size() {
+        return properties.size();
+    }
 
     /**
-     * Map to hold analytics properties.
+     * get returns a specific {@link Property} by its name.
+     *
+     * @param propertyName The name of the property to retrieve
+     * @return The corresponding {@link Property}
      */
-    @SuppressWarnings("VisibilityModifier")
-    protected final Map<String, Property<?>> properties;
+    @NonNull
+    public Property<?> get(@NonNull String propertyName) {
+        return properties.get(propertyName);
```

Comment:
If I pass a non-existing property name, this would currently return `null`, but the flag says `@NonNull`. I think we need some logic like:
```
Property<?> property = properties.get(propertyName);
if (property == null) {
   // there was no match ... throw an exception?
}
return property;
```

Reply:
I suppose we could also remove the @NonNull and embrace the null-ness as this method is mainly used for test purposes. I gave it a shot by raising `NoSuchElementException`.